plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/alarms/Alarm.java

```diff
@@ -0,0 +1,134 @@
+package eu.arrowhead.core.plantdescriptionengine.alarms;
+
+import eu.arrowhead.core.plantdescriptionengine.providedservices.pde_monitor.dto.PdeAlarmDto;
+
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Internal representation of a PDE Alarm.
+ */
+public class Alarm {
+
+    // Integer for storing the next alarm ID to be used:
+    private static final AtomicInteger nextId = new AtomicInteger();
+    private static final String unknownId = "Unknown";
+
+    private final int id;
+    private final String systemName;
+    private final String systemId;
+    private final AlarmCause cause;
+
+    private final Map<String, String> metadata;
+    private final Instant raisedAt;
+    private boolean acknowledged;
+    private Instant updatedAt;
+    private Instant clearedAt;
+    private Instant acknowledgedAt;
+
+    Alarm(final String systemId, final String systemName, final Map<String, String> metadata, final AlarmCause cause) {
+
+        Objects.requireNonNull(cause, "Expected an alarm cause.");
+        Objects.requireNonNull(metadata, "Expected metadata.");
+
+        this.systemId = systemId;
+        this.systemName = systemName;
+        this.metadata = metadata;
+        this.cause = cause;
+
+        id = nextId.getAndIncrement();
+        acknowledged = false;
+        raisedAt = Instant.now();
+        updatedAt = raisedAt;
+    }
+
+    /**
+     * @param systemId   ID of a system or null.
+     * @param systemName Name of a system or null.
+     * @param metadata   A system's metadata, or null.
+     * @param cause      An alarm cause.
+     * @return True if the provided arguments match the data stored in this
+     * alarm instance.
+     */
+    public boolean matches(final String systemId, final String systemName, final Map<String, String> metadata, final AlarmCause cause) {
+        Objects.requireNonNull(cause, "Expected alarm cause.");
+        final Map<String, String> nonNullMetadata = metadata == null ? Collections.emptyMap() : metadata;
+        return cause == this.cause &&
+            Objects.equals(systemId, this.systemId) &&
+            Objects.equals(systemName, this.systemName) &&
```

Comment:
I don't see where the systemName parameter comes from, but if it comes from the Service Registry, then the previously mentioned normalization can cause issues here.

Reply:
System names are now validated at the time they enter the PDE.